File: Week 1/Union Find/quickfind.py
# QuickFind(Union Find) Data Structure
# It has O(n^2) time constant which is not ideal

class QuickFindUF:
    def __init__(self, N):
        self.id = [i for i in range(N)]

    def connected(self, p, q):
        return self.id[p] == self.id[q]

    def union(self, p, q):
        index = 0
        pid = self.id[p]
        while(index < len(self.id)):
            if (self.id[index] == pid):
                self.id[index] = self.id[q]
            index += 1

victim = QuickFindUF(10)
count = 0
while (count < len(victim.id)):
    print(victim.id[count], end=" ")
    count += 1
print()

victim.union(4, 3)
victim.union(3, 8)
victim.union(6, 5)
victim.union(9, 4)
victim.union(2, 1)

count = 0
while (count < len(victim.id)):
    print(victim.id[count], end=" ")
    count += 1
print()
print(victim.connected(8,9))
print(victim.connected(5,0))

File: Week 1/Union Find/quickfind.cpp
// QuickFind (Union Find) data structure. It has time constant of O(n^2)
#include <iostream>
#include <vector>
using namespace std;

class QuickFindUf
{
public:
    vector<int> array;
    QuickFindUf(int N) {
        for(int index = 0; index < N; index++)
            array.push_back(index);
    }

    int Connected(int p, int q) {
        return array.at(p) == array.at(q);
    }

    void Union(int p, int q) {
        int pid = array.at(p);
        for(int index = 0; index < array.size(); index++) 
            if(array.at(index) == pid)
                array.at(index) = array.at(q);
    }
    void display() {
        for(int index = 0; index < array.size(); index++)
            printf("%d ", array.at(index));
        printf("\n");
    }
};

int main()
{
    QuickFindUf victim(10);
    victim.display();
    victim.Union(4, 3);
    victim.Union(3, 8);
    victim.Union(6, 5);
    victim.Union(9, 4);
    victim.Union(2, 1);
    victim.display();
    printf("victim.connected(8,9): %d\n", victim.Connected(8, 9));
    printf("victim.connected(5,0): %d\n", victim.Connected(5, 0));
}